File: sdks/android/src/main/kotlin/com/adxplatform/sdk/internal/BidResponseParser.kt
package com.taskirx.sdk.internal

import com.taskirx.sdk.config.AdFormat
import com.taskirx.sdk.models.AdResponse
import com.taskirx.sdk.models.BidResponse
import org.json.JSONObject

/**
 * Parses OpenRTB bid responses into AdResponse objects
 */
internal object BidResponseParser {
    
    fun parse(bidResponse: BidResponse, adFormat: AdFormat): AdResponse? {
        val seatBid = bidResponse.seatBids?.firstOrNull() ?: return null
        val bid = seatBid.bids.firstOrNull() ?: return null
        
        return when (adFormat) {
            AdFormat.BANNER, AdFormat.IN_APP_BANNER -> parseBanner(bid, bidResponse.bidId ?: bid.id)
            AdFormat.VIDEO, AdFormat.IN_APP_VIDEO, AdFormat.REWARDED_VIDEO -> parseVideo(bid, bidResponse.bidId ?: bid.id)
            AdFormat.NATIVE, AdFormat.IN_APP_NATIVE -> parseNative(bid, bidResponse.bidId ?: bid.id)
            AdFormat.INTERSTITIAL -> parseBanner(bid, bidResponse.bidId ?: bid.id)
        }
    }
    
    private fun parseBanner(bid: com.taskirx.sdk.models.Bid, bidId: String): AdResponse {
        // Ad markup should contain image URL or HTML
        val markup = bid.adMarkup
        
        // Try to extract image URL from markup
        val imageUrl = extractImageUrl(markup) ?: markup
        val clickUrl = extractClickUrl(markup)
        
        return AdResponse(
            bidId = bidId,
            impressionId = bid.impressionId,
            price = bid.price,
            adFormat = AdFormat.BANNER,
            imageUrl = imageUrl,
            clickUrl = clickUrl,
            impressionUrl = bid.noticeUrl,
            width = bid.width,
            height = bid.height,
            extensions = bid.extensions
        )
    }
    
    private fun parseVideo(bid: com.taskirx.sdk.models.Bid, bidId: String): AdResponse {
        // Ad markup should contain VAST XML or video URL
        val videoUrl = extractVideoUrl(bid.adMarkup) ?: bid.adMarkup
        val clickUrl = extractClickUrl(bid.adMarkup)
        
        return AdResponse(
            bidId = bidId,
            impressionId = bid.impressionId,
            price = bid.price,
            adFormat = AdFormat.VIDEO,
            videoUrl = videoUrl,
            clickUrl = clickUrl,
            impressionUrl = bid.noticeUrl,
            extensions = bid.extensions
        )
    }
    
    private fun parseNative(bid: com.taskirx.sdk.models.Bid, bidId: String): AdResponse {
        // Parse native ad markup (JSON)
        try {
            val json = JSONObject(bid.adMarkup)
            val native = json.optJSONObject("native")
            
            var title: String? = null
            var description: String? = null
            var imageUrl: String? = null
            var iconUrl: String? = null
            var clickUrl: String? = null
            var callToAction: String? = null
            var sponsoredBy: String? = null
            
            // Parse assets
            val assets = native?.optJSONArray("assets")
            if (assets != null) {
                for (i in 0 until assets.length()) {
                    val asset = assets.getJSONObject(i)
                    val id = asset.optInt("id")
                    
                    when (id) {
                        1 -> title = asset.optJSONObject("title")?.optString("text")
                        2 -> imageUrl = asset.optJSONObject("img")?.optString("url")
                        3 -> description = asset.optJSONObject("data")?.optString("value")
                    }
                }
            }
            
            // Parse link
            val link = native?.optJSONObject("link")
            clickUrl = link?.optString("url")
            
            return AdResponse(
                bidId = bidId,
                impressionId = bid.impressionId,
                price = bid.price,
                adFormat = AdFormat.NATIVE,
                title = title,
                description = description,
                imageUrl = imageUrl,
                iconUrl = iconUrl,
                clickUrl = clickUrl,
                callToAction = callToAction ?: "Learn More",
                sponsoredBy = sponsoredBy,
                impressionUrl = bid.noticeUrl,
                extensions = bid.extensions
            )
        } catch (e: Exception) {
            throw IllegalArgumentException("Failed to parse native ad response", e)
        }
    }
    
    private fun extractImageUrl(markup: String): String? {
        // Extract image URL from HTML
        val imgRegex = """<img[^>]+src=["']([^"']+)["']""".toRegex()
        return imgRegex.find(markup)?.groupValues?.get(1)
    }
    
    private fun extractClickUrl(markup: String): String? {
        // Extract click URL from HTML
        val hrefRegex = """<a[^>]+href=["']([^"']+)["']""".toRegex()
        return hrefRegex.find(markup)?.groupValues?.get(1)
    }
    
    private fun extractVideoUrl(markup: String): String? {
        // Extract video URL from VAST XML or markup
        val videoRegex = """<MediaFile[^>]*>([^<]+)</MediaFile>""".toRegex()
        val match = videoRegex.find(markup)?.groupValues?.get(1)?.trim()
        
        // If not VAST, check for direct video URL
        return match ?: if (markup.contains(".mp4") || markup.contains(".webm")) {
            markup.trim()
        } else {
            null
        }
    }
}

File: sdks/ios/Sources/AdxSDK/BidResponseParser.swift
import Foundation

internal class BidResponseParser {
    
    static func parse(bidResponse: BidResponse, adFormat: AdFormat) -> AdResponse? {
        guard let seatBid = bidResponse.seatbid?.first,
              let bid = seatBid.bid.first else {
            return nil
        }
        
        switch adFormat {
        case .banner, .interstitial:
            return parseBanner(bid: bid, bidId: bidResponse.bidid ?? bid.id)
        case .video, .rewardedVideo:
            return parseVideo(bid: bid, bidId: bidResponse.bidid ?? bid.id)
        case .native:
            return parseNative(bid: bid, bidId: bidResponse.bidid ?? bid.id)
        }
    }
    
    private static func parseBanner(bid: Bid, bidId: String) -> AdResponse {
        let markup = bid.adm
        let imageURL = extractImageURL(from: markup)
        let clickURL = extractClickURL(from: markup)
        
        return AdResponse(
            bidId: bidId,
            impressionId: bid.impid,
            price: bid.price,
            adFormat: .banner,
            imageURL: imageURL,
            clickURL: clickURL,
            impressionURL: bid.nurl.flatMap { URL(string: $0) },
            videoURL: nil,
            title: nil,
            description: nil,
            callToAction: nil,
            iconURL: nil,
            sponsoredBy: nil,
            width: bid.w,
            height: bid.h
        )
    }
    
    private static func parseVideo(bid: Bid, bidId: String) -> AdResponse {
        let markup = bid.adm
        let videoURL = extractVideoURL(from: markup)
        let clickURL = extractClickURL(from: markup)
        
        return AdResponse(
            bidId: bidId,
            impressionId: bid.impid,
            price: bid.price,
            adFormat: .video,
            imageURL: nil,
            clickURL: clickURL,
            impressionURL: bid.nurl.flatMap { URL(string: $0) },
            videoURL: videoURL,
            title: nil,
            description: nil,
            callToAction: nil,
            iconURL: nil,
            sponsoredBy: nil,
            width: bid.w,
            height: bid.h
        )
    }
    
    private static func parseNative(bid: Bid, bidId: String) -> AdResponse {
        guard let data = bid.adm.data(using: .utf8),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let native = json["native"] as? [String: Any] else {
            return AdResponse(
                bidId: bidId,
                impressionId: bid.impid,
                price: bid.price,
                adFormat: .native,
                imageURL: nil,
                clickURL: nil,
                impressionURL: bid.nurl.flatMap { URL(string: $0) },
                videoURL: nil,
                title: "Ad",
                description: "Sponsored content",
                callToAction: "Learn More",
                iconURL: nil,
                sponsoredBy: nil,
                width: bid.w,
                height: bid.h
            )
        }
        
        var title: String?
        var description: String?
        var imageURL: URL?
        var iconURL: URL?
        var clickURL: URL?
        var callToAction: String?
        
        // Parse assets
        if let assets = native["assets"] as? [[String: Any]] {
            for asset in assets {
                let id = asset["id"] as? Int ?? 0
                
                switch id {
                case 1: // Title
                    if let titleData = asset["title"] as? [String: Any],
                       let text = titleData["text"] as? String {
                        title = text
                    }
                case 2: // Main image
                    if let imgData = asset["img"] as? [String: Any],
                       let urlString = imgData["url"] as? String {
                        imageURL = URL(string: urlString)
                    }
                case 3: // Description
                    if let dataObj = asset["data"] as? [String: Any],
                       let value = dataObj["value"] as? String {
                        description = value
                    }
                default:
                    break
                }
            }
        }
        
        // Parse link
        if let link = native["link"] as? [String: Any],
           let urlString = link["url"] as? String {
            clickURL = URL(string: urlString)
        }
        
        return AdResponse(
            bidId: bidId,
            impressionId: bid.impid,
            price: bid.price,
            adFormat: .native,
            imageURL: imageURL,
            clickURL: clickURL,
            impressionURL: bid.nurl.flatMap { URL(string: $0) },
            videoURL: nil,
            title: title,
            description: description,
            callToAction: callToAction ?? "Learn More",
            iconURL: iconURL,
            sponsoredBy: nil,
            width: bid.w,
            height: bid.h
        )
    }
    
    // MARK: - Helper Methods
    
    private static func extractImageURL(from markup: String) -> URL? {
        // Extract image URL from HTML
        let pattern = #"<img[^>]+src=["']([^"']+)["']"#
        guard let regex = try? NSRegularExpression(pattern: pattern),
              let match = regex.firstMatch(in: markup, range: NSRange(markup.startIndex..., in: markup)),
              match.numberOfRanges > 1,
              let range = Range(match.range(at: 1), in: markup) else {
            return URL(string: markup) // Fallback: treat as direct URL
        }
        
        return URL(string: String(markup[range]))
    }
    
    private static func extractClickURL(from markup: String) -> URL? {
        // Extract click URL from HTML
        let pattern = #"<a[^>]+href=["']([^"']+)["']"#
        guard let regex = try? NSRegularExpression(pattern: pattern),
              let match = regex.firstMatch(in: markup, range: NSRange(markup.startIndex..., in: markup)),
              match.numberOfRanges > 1,
              let range = Range(match.range(at: 1), in: markup) else {
            return nil
        }
        
        return URL(string: String(markup[range]))
    }
    
    private static func extractVideoURL(from markup: String) -> URL? {
        // Extract video URL from VAST XML or direct URL
        let pattern = #"<MediaFile[^>]*>([^<]+)</MediaFile>"#
        guard let regex = try? NSRegularExpression(pattern: pattern),
              let match = regex.firstMatch(in: markup, range: NSRange(markup.startIndex..., in: markup)),
              match.numberOfRanges > 1,
              let range = Range(match.range(at: 1), in: markup) else {
            // Try as direct URL
            if markup.contains(".mp4") || markup.contains(".mov") {
                return URL(string: markup.trimmingCharacters(in: .whitespacesAndNewlines))
            }
            return nil
        }
        
        let urlString = String(markup[range]).trimmingCharacters(in: .whitespacesAndNewlines)
        return URL(string: urlString)
    }
}
